src/app/chip-zcl-zpro-codec.h

```diff
@@ -115,29 +89,62 @@ void printApsFrame(EmberApsFrame * frame);
  */
 uint16_t encodeApsFrame(uint8_t * buffer, uint16_t buf_length, EmberApsFrame * apsFrame);
 
-/**
- * Identify cluster commands
- * */
+#ifdef __cplusplus
+}
+#endif
+
+uint16_t encodeClusterSpecificCommandHeader(chip::BufBound & buf, uint8_t endpointId, uint16_t clusterId, uint8_t commandId);
+
+uint16_t encodeGlobalCommandHeader(chip::BufBound & buf, uint8_t endpointId, uint16_t clusterId, uint8_t commandIdId);
+
+template <typename... Args>
+void encodeCommand(chip::BufBound & buf, Args... args)
+{}
+
+template <typename... Args>
+void encodeCommand(chip::BufBound & buf, uint8_t value, Args... args)
+{
+    buf.Put(value);
+    encodeCommand(buf, args...);
+}
+
+template <typename... Args>
+void encodeCommand(chip::BufBound & buf, uint16_t value, Args... args)
+{
+    buf.PutLE16(value);
+    encodeCommand(buf, args...);
+}
 
 /**
- * @brief Encode an identify query command for the identify cluster
- * @param buffer Buffer to encode into
- * @param buf_length Length of buffer
- * @param destination_endpoint destination endpoint
- * */
-uint16_t encodeIdentifyQueryCommand(uint8_t * buffer, uint16_t buf_length, uint8_t destination_endpoint);
+ * @brief Encode a Cluster Specific command for the given cluster and the given values.
+ */
+template <typename... Args>
+uint16_t encodeCommand(uint8_t * buffer, uint16_t bufferLen, uint8_t endpointId, uint16_t clusterId, uint8_t commandId,
+                       Args... args)
+{
+    chip::BufBound buf = chip::BufBound(buffer, bufferLen);
+    if (encodeClusterSpecificCommandHeader(buf, endpointId, clusterId, commandId))
+    {
+        encodeCommand(buf, args...);
+    }
+
+    return buf.Fit() && chip::CanCastTo<uint16_t>(buf.Written()) ? static_cast<uint16_t>(buf.Written()) : 0;
+}
 
 /**
- * @brief Encode an identify command for the identify cluster
- * @param buffer Buffer to encode into
- * @param buf_length Length of buffer
- * @param destination_endpoint destination endpoint
- * @param duration The duration for which the device will identify itself
- * */
-uint16_t encodeIdentifyCommand(uint8_t * buffer, uint16_t buf_length, uint8_t destination_endpoint, uint16_t duration);
+ * @brief Encode a Global command for the given cluster and the given values.
+ */
+template <typename... Args>
+uint16_t encodeGlobalCommand(uint8_t * buffer, uint16_t bufferLen, uint8_t endpointId, uint16_t clusterId, uint8_t commandId,
+                             Args... args)
+{
+    chip::BufBound buf = chip::BufBound(buffer, bufferLen);
+    if (encodeGlobalCommandHeader(buf, endpointId, clusterId, commandId))
+    {
+        encodeCommand(buf, args...);
+    }
 
-#ifdef __cplusplus
+    return buf.Fit() && chip::CanCastTo<uint16_t>(buf.Written()) ? static_cast<uint16_t>(buf.Written()) : 0;
```

Comment:
no reason that BufBound can't move to uint16_t for lengths now...  everything else is in shorts